{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "867c117f",
   "metadata": {},
   "source": [
    "# MY MODEL (P11018)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7678081d",
   "metadata": {},
   "source": [
    "## 1. Find homologs with HMM:\n",
    "\n",
    "```psiblast -query P11018.fa -db /mnt/NFS_UPF/soft/databases/blastdat/pdb_seq -num_iterations 5 -out P11018_pdb.out```\n",
    "\n",
    "And will obtain this:\n",
    "\n",
    "````\n",
    "Query= P11018\n",
    "\n",
    "Length=319\n",
    "                                                                      Score     E\n",
    "Sequences producing significant alignments:                          (Bits)  Value\n",
    "\n",
    "1mee_A mol:protein length:275 MESENTERICOPEPTIDASE                    226     2e-72\n",
    "1scj_A mol:protein length:275 SUBTILISIN E                            223     3e-71\n",
    "1yjc_A mol:protein length:275 SUBTILISIN 8397+1                       218     2e-69\n",
    "1yjb_A mol:protein length:275 SUBTILISIN 8397+1                       218     2e-69\n",
    "1yja_A mol:protein length:275 SUBTILISIN 8397+1                       218     2e-69\n",
    "1sbh_A mol:protein length:275 SUBTILISIN 8397+1                       218    \n",
    "````\n",
    "\n",
    "With this I know that 1meeA, 1scjA and 1yjc are the ones I will be using\n",
    "\n",
    "### 1.1. Download the fam profile with PFAMa db\n",
    "\n",
    "Busquem l'HMM de la família utilitzant la base de dades de PFAMa. \n",
    "\n",
    "````\n",
    "hmmpfam2 /mnt/NFS_UPF/soft/databases/pfam-2/Pfam_ls P11018.fa > P11018_pfam.out\n",
    "````\n",
    "\n",
    "Obtinc:\n",
    "\n",
    "````\n",
    "Scores for sequence family classification (score includes all domains):\n",
    "Model           Description                             Score    E-value  N \n",
    "--------        -----------                             -----    ------- ---\n",
    "Peptidase_S8    Subtilase family                        411.9   1.1e-120   1\n",
    "LacAB_rpiB      Ribose/Galactose Isomerase              -96.7          1   1\n",
    "ETF             Electron transfer flavoprotein domain   -56.0          7   1\n",
    "ABC_sub_bind    ABC transporter substrate binding pro  -158.2        9.2   1\n",
    "````\n",
    "\n",
    "Ara sabem que pertanyen a la família de la Peptidase_S8! Ens servirà epr extreure el seu HMM:\n",
    "\n",
    "With ```hmmfetch```, because I already know that P11018 comes from the globin superfamily.\n",
    "\n",
    "````\n",
    "hmmfetch /mnt/NFS_UPF/soft/databases/pfam-3/Pfam-A.hmm Peptidase_S8 > Peptidase_S8.hmm\n",
    "````\n",
    "\n",
    "### 1.2. Once have a specific HMM model, use ```hmmsearch``` against a sequence db (PDB; Unirpot) to find the homologs\n",
    "        \n",
    "````\n",
    "hmmsearch Peptidase_S8.hmm /mnt/NFS_UPF/soft/databases/blastdat/pdb_seq > Peptidase_S8_pdb.out\n",
    "````\n",
    "\n",
    "From this file, I will find the homologous seq. I did it before with psiblast, but the results now are quite different...\n",
    "\n",
    "So as a templates, I will choose: 1sbhA, 1yjaA and iyjbA\n",
    "\n",
    "\n",
    "Un cop triats els templates, anar al fitxer original /mnt/NFS_UPF/soft/databases/blastdat/pdb_seq i copiar les seves seqüències FASTA a un fitxer nou (anomena'l templates.fa). To obtain the .fa files for each of this two sequences: \n",
    "\n",
    "````\n",
    "cp /mnt/NFS_UPF/soft/databases/pdb/pdb/data/structures/divided/pdb/sb/pdb1sbh.ent.gz .\n",
    "cp /mnt/NFS_UPF/soft/databases/pdb/pdb/data/structures/divided/pdb/yj/pdb1yja.ent.gz .\n",
    "cp /mnt/NFS_UPF/soft/databases/pdb/pdb/data/structures/divided/pdb/yj/pdb1yjb.ent.gz .\n",
    "\n",
    "gunzip *gz\n",
    "\n",
    "cp pdb1sbh.ent 1sbh.pdb\n",
    "cp pdb1yja.ent 1yja.pdb\n",
    "cp pdb1yjb.ent 1yjb.pdb\n",
    "\n",
    "PDBtoSplitChain.pl -i 1sbh.pdb -o 1sbh\n",
    "PDBtoSplitChain.pl -i 1yja.pdb -o 1yja\n",
    "PDBtoSplitChain.pl -i 1yjb.pdb -o 1yjb\n",
    "````\n",
    "With the PDBtoSplitChain I have obtained chains A. \n",
    "\n",
    "````\n",
    "cat P11018.fa > alignment.fa\n",
    "cat 1sbhA.fa >> alignment.fa\n",
    "cat 1yjaA.fa >> alignment.fa\n",
    "cat 1yjbA.fa >> alignment.fa\n",
    "````\n",
    "\n",
    "\n",
    "## 2. Align the ```hmmfetch``` results with the hmm profile using ```hmmalign```\n",
    "\n",
    "Here I want P11018 to align it with other homologs obtained from the ```hmmsearch``` . A diferència d'un aliniament seqüència-seqüència normal (com faria ClustalW), aquí estic utilitzant una \"plantilla intel·ligent\" (el model HMM de Pfam) per col·locar cada aminoàcid al seu lloc correcte segons l'evolució de la família de les globines. El programa mira el model HMM, que ja sap quines posicions de la globina són hèlixs, quines són girs i on és probable que hi hagi una inserció o una deleció (gaps). Agafa la teva P11018 i les templates (1abwA, 1o1nA) i les projecta sobre el model una per una. Com que totes s'han alineat contra el mateix motlle, el resultat és que estan alineades entre elles de la manera més precisa possible des d'un punt de vista biològic.\n",
    "\n",
    "````\n",
    "hmmalign Peptidase_S8.hmm alignment.fa > final_alignment.sto\n",
    "````\n",
    "\n",
    "\n",
    "## 3. Convert the sto format into alignment one (clustal)\n",
    "\n",
    "````\n",
    "aconvertMod2.pl -in h -out c < final_alignment.sto > final_alignment.clu\n",
    "````\n",
    "\n",
    "## 3.1. Output of the alignment in clustal format\n",
    "\n",
    "````\n",
    "CLUSTAL W(1.60) multiple sequence alignment\n",
    "\n",
    "P11018  mngeirlipyvtneqimdvnelpegikvikapemwakgvkgknikvavldtgcdtshpdl\n",
    "1abwA   --------------mlspadktnvkaawgkvgahageygaealermflsfpttktyfphf\n",
    "1o1nA   ---------------mlspadktnvkaawgkvgahageygaeawermflsfpttktyfph\n",
    "\n",
    "P11018  knqiiggknftdddggkedaisdynghgthvagtiaandsnggiagvapeasllivkvlg\n",
    "1abwA   dlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvnfkllshcll\n",
    "1o1nA   fdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvnfkllshcl\n",
    "\n",
    "P11018  gengsgqyewiinginyaveqkvdiismslggpsdvpelke-------------------\n",
    "1abwA   vtlaahlpaeftpavhasldkflasvstvltskyrgvlspaD------------------\n",
    "1o1nA   lvtlaahlpaeftpavhasldkflasvstvltskyrvlspaD------------------\n",
    "\n",
    "P11018  -------------------------------------------------A----------\n",
    "1abwA   ------------------------------------------------------------\n",
    "1o1nA   ------------------------------------------------------------\n",
    "\n",
    "P11018  -----------------------------vknavkngvlvvcaagnegdgderteelsyp\n",
    "1abwA   -----------------------------ktnvkaawgkvgahageygaealermflsfp\n",
    "1o1nA   -----------------------------ktnvkaawgkvgahageygaeawermflsfp\n",
    "\n",
    "P11018  aayneviavgsvsvarelsefsnankeidlvapgenilstlpnkkygkltgtsmaaphvs\n",
    "1abwA   ttktyfphfdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvn\n",
    "1o1nA   ttktyfphfdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvn\n",
    "\n",
    "P11018  galaliksyeeesfqrklsesevfaqlirrtlpldiaktlagngflyltapdelaekaeq\n",
    "1abwA   fkllshcllvtlaahlpaeftpavhasldkflasvstvltskyr----------------\n",
    "1o1nA   fkllshcllvtlaahlpaeftpavhasldkflasvstvltskyr----------------\n",
    "\n",
    "P11018  shlltl\n",
    "1abwA   ------\n",
    "1o1nA   ------\n",
    "````"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e094d151",
   "metadata": {},
   "source": [
    "## 4. MODELLER\n",
    "\n",
    "Ara fem servir el programa MODELLER per construir el nostre model. Modeller és una implementació automàtica d'una aproximació de modelatge comparatiu que segueix restriccions espacials. Funciona de tal manera que primer, les estructures 3D de les templates s'alinien amb la target, després s'apliquen les restriccions espacials (distància entre Calpha, cadenes laterals, angles...) que es transfereixen de les templates a la target, finalment, s'obté el model 3D. Aquest complirà al màxim de restriccions possibles. \n",
    "\n",
    "Modeller només llegirà l'alineament si aquest està en format pir. El programa ```aconvertMod2``` permet fer aquest canvi de format. L'output és un alineament amb tants blocs com seq i cada bloc té un encapçalament i dos línies amb etqiuetes. Aquestes etiquetes hauràn de ser les mateixes que les que posarem a l'input file. \n",
    "\n",
    "````\n",
    "aconvertMod2.pl -in c -out p < final_alignment.clu > final_alignment.pir\n",
    "````\n",
    "\n",
    "### 4.1. Modify ```modeling.py``` file\n",
    "\n",
    "Aquest file conté els paràmetres que necessita Modeller i les opcions que volem fer servir per a construir el meu model 3D. \n",
    "\n",
    "- ```alnfile``` on es posa el nom de l'alinimanet en format pir, a partir del qual es construeix els models. Entre cometes!\n",
    "\n",
    "- ```knowns``` on es posa el nom dels templates utilitzats. Han de tenir el amteix nom que en l'aliniament pir. Entre cometes i separat entre cometes!\n",
    "\n",
    "- ```sequence``` nom del target (en aquest cas P11018) que també ha d'apareixer a l'aliniment pir. \n",
    "\n",
    "Els pdb han d'estar al mateix directori\n",
    "\n",
    "- ```a.ending_model = 2``` perquè es vol dos models, tot i que es podria fer tants models com vulgui perquè hi ha més d'una solució per a cada restricció espacial. \n",
    "\n",
    "- ```a.make()``` NO ES TOCA! Serveix per fer homology modeling de la manera més simple. \n",
    "\n",
    "## 4.2. Run modeling.py\n",
    "\n",
    "Un cop canviat els paràmetres, correm el file .py:\n",
    "\n",
    "````\n",
    "modpy.sh python3 modeling.py\n",
    "````\n",
    "\n",
    "Com a resultat, obtindré dos models:\n",
    "\n",
    "````\n",
    ">> Summary of successfully produced models:\n",
    "Filename                          molpdf\n",
    "----------------------------------------\n",
    "P11018.B99990001.pdb          7236.42578\n",
    "P11018.B99990002.pdb          7223.64941\n",
    "````\n",
    "\n",
    "Per més comoditat, els canviaré el nom:\n",
    "\n",
    "````\n",
    "mv P11018.B99990001.pdb model1.pdb\n",
    "mv P11018.B99990002.pdb model2.pdb\n",
    "````\n",
    "## 4.3. Visualization on chimera\n",
    "Visualització amb chimera!!\n",
    "\n",
    "````\n",
    "chimera model1.pdb model2.pdb 1sbhA.pdb 1yjaA.pdb 1yjbA.pdb\n",
    "````\n",
    "\n",
    "I superimposem. Es veu que no acaba de coincidir massa i això pot ser degut a que hi ha gaps al principi i/o al final de l'aliniement. Llavors es recomanable construir nous models sense auqest segment, esborrant certs aminaocids de l'arxiu fasta. \n",
    "\n",
    "## 4.4. Modify alignment to enhance the superimposition\n",
    "\n",
    "__⚠️ EL ALINIAMENT SEGÜENT NO CORRESPON AL ALINIAMENT ORIGINAL, EM VAIG EQUIVOCAR I NOMÉS MOSTRO UN EXEMPLE PER ACLARIR EL QUE ES VOL FER!!__\n",
    "Abans:\n",
    "\n",
    "````\n",
    "CLUSTAL W(1.60) multiple sequence alignment\n",
    "\n",
    "P11018  mngeirlipyvtneqimdvnelpegikvikapemwakgvkgknikvavldtgcdtshpdl\n",
    "1abwA   --------------mlspadktnvkaawgkvgahageygaealermflsfpttktyfphf\n",
    "1o1nA   ---------------mlspadktnvkaawgkvgahageygaeawermflsfpttktyfph\n",
    "\n",
    "P11018  knqiiggknftdddggkedaisdynghgthvagtiaandsnggiagvapeasllivkvlg\n",
    "1abwA   dlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvnfkllshcll\n",
    "1o1nA   fdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvnfkllshcl\n",
    "\n",
    "P11018  gengsgqyewiinginyaveqkvdiismslggpsdvpelke-------------------\n",
    "1abwA   vtlaahlpaeftpavhasldkflasvstvltskyrgvlspaD------------------\n",
    "1o1nA   lvtlaahlpaeftpavhasldkflasvstvltskyrvlspaD------------------\n",
    "\n",
    "P11018  -------------------------------------------------A----------\n",
    "1abwA   ------------------------------------------------------------\n",
    "1o1nA   ------------------------------------------------------------\n",
    "\n",
    "P11018  -----------------------------vknavkngvlvvcaagnegdgderteelsyp\n",
    "1abwA   -----------------------------ktnvkaawgkvgahageygaealermflsfp\n",
    "1o1nA   -----------------------------ktnvkaawgkvgahageygaeawermflsfp\n",
    "\n",
    "P11018  aayneviavgsvsvarelsefsnankeidlvapgenilstlpnkkygkltgtsmaaphvs\n",
    "1abwA   ttktyfphfdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvn\n",
    "1o1nA   ttktyfphfdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvn\n",
    "\n",
    "P11018  galaliksyeeesfqrklsesevfaqlirrtlpldiaktlagngflyltapdelaekaeq\n",
    "1abwA   fkllshcllvtlaahlpaeftpavhasldkflasvstvltskyr----------------\n",
    "1o1nA   fkllshcllvtlaahlpaeftpavhasldkflasvstvltskyr----------------\n",
    "\n",
    "P11018  shlltl\n",
    "1abwA   ------\n",
    "1o1nA   ------\n",
    "````\n",
    "Ara:\n",
    "````\n",
    "CLUSTAL W(1.60) multiple sequence alignment\n",
    "\n",
    "P11018  qimdvnelpegikvikapemwakgvkgknikvavldtgcdtshpdl\n",
    "1abwA   mlspadktnvkaawgkvgahageygaealermflsfpttktyfphf\n",
    "1o1nA   mlspadktnvkaawgkvgahageygaeawermflsfpttktyfph\n",
    "\n",
    "P11018  knqiiggknftdddggkedaisdynghgthvagtiaandsnggiagvapeasllivkvlg\n",
    "1abwA   dlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvnfkllshcll\n",
    "1o1nA   fdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvnfkllshcl\n",
    "\n",
    "P11018  gengsgqyewiinginyaveqkvdiismslggpsdvpelke-------------------\n",
    "1abwA   vtlaahlpaeftpavhasldkflasvstvltskyrgvlspaD------------------\n",
    "1o1nA   lvtlaahlpaeftpavhasldkflasvstvltskyrvlspaD------------------\n",
    "\n",
    "P11018  -------------------------------------------------A----------\n",
    "1abwA   ------------------------------------------------------------\n",
    "1o1nA   ------------------------------------------------------------\n",
    "\n",
    "P11018  -----------------------------vknavkngvlvvcaagnegdgderteelsyp\n",
    "1abwA   -----------------------------ktnvkaawgkvgahageygaealermflsfp\n",
    "1o1nA   -----------------------------ktnvkaawgkvgahageygaeawermflsfp\n",
    "\n",
    "P11018  aayneviavgsvsvarelsefsnankeidlvapgenilstlpnkkygkltgtsmaaphvs\n",
    "1abwA   ttktyfphfdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvn\n",
    "1o1nA   ttktyfphfdlshgsaqvkghgkkvadaltnavahvddmpnalsalsdlhahklrvdpvn\n",
    "\n",
    "P11018  galaliksyeeesfqrklsesevfaqlirrtlpldiaktlagng\n",
    "1abwA   fkllshcllvtlaahlpaeftpavhasldkflasvstvltskyr\n",
    "1o1nA   fkllshcllvtlaahlpaeftpavhasldkflasvstvltskyr\n",
    "````\n",
    "Ara, tornem a fer:\n",
    "\n",
    "````\n",
    "aconvertMod2.pl -in c -out p < final_alignment.clu > final_alignment.pir\n",
    "modpy.sh python3 modeling.py\n",
    "\n",
    "mv P11018.B99990001.pdb model1_withouttail.pdb\n",
    "mv P11018.B99990002.pdb model2_withouttail.pdb\n",
    "\n",
    "chimera model1_withouttail.pdb model2_withouttail.pdb 1sbhA.pdb 1yjaA.pdb 1yjbA.pdb\n",
    "````\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
